Proxy

wprzytula opened the pull request:
This PR adds proxy layer between driver and cluster for testing driver in unfavourable conditions. The proxy is added in a separate crate. Its capabilities include:
            - verbose feedback: connections created and frames sent,
            - dropping connections...,
            - dropping, delaying and forging frames...,
            - ...based on configurable rules activated on given conditions,
            - being integrated into driver's integration tests.

As proxy can't live without address translation, I reverted back the reversal of my previous commit, but also made address translation leave addresses provided as contact points untouched.

Moreover, I added two simple tests for retries and speculative executions. Having found a bug in driver thanks to them, I fixed it. 

## Pre-review checklist
<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [ ] I added appropriate `Fixes:` annotations to PR description.

lcmgh commented:
Is there meanwhile another way to get a connection with an http proxy working when it is corporate blocked?

wprzytula commented:
> 

Could you elaborate?
This PR (the whole _proxy_ thing) is just a layer that would intercept, log and possibly mutate all frames between the driver and the cluster, mainly for debug/test purposes.

lcmgh commented:
I am working in a corporate environment and the direct connection to the AWS Public Cassandra Endpoint does only work with help of a proxy. I'm not sure how I can utilize a http proxy (which I may can transform to a Socks proxy) to connect to the AWS endpoint.

wprzytula commented:
I've increased commits granularity, the proxy is ready for review.

cvybhu commented:
> I am working in a corporate environment and the direct connection to the AWS Public Cassandra Endpoint does only work with help of a proxy. I'm not sure how I can utilize a http proxy (which I may can transform to a Socks proxy) to connect to the AWS endpoint.

@lcmgh This proxy is intended for sniffing network requests and seeing what the driver does. It doesn't tunnel requests like a SOCKS proxy. It's more like wireshark, used mostly for debugging and testing.

It looks like you could maybe use address translation. It allows you to specify a different address for database nodes. It could be pointed to some kind of proxy instead of the real node address.

It's a very new feature, it isn't on crates.io and there's no documentation yet, but you can use driver version from the `main` branch to test it. 
An example would be:
* Specifying address translation map: https://github.com/scylladb/scylla-rust-driver/blob/6c74cd2cfc33e04b8cc26c3f08d41dc98f7b0ecf/scylla/tests/retries.rs#L100
* Telling `Session` to translate addresses: https://github.com/scylladb/scylla-rust-driver/blob/6c74cd2cfc33e04b8cc26c3f08d41dc98f7b0ecf/scylla/tests/retries.rs#L221

If you have any more questions please open an issue. Describe your situation in detail (existing networks, proxies, etc.) and we might be able to find some solution. Let's keep comments in this PR focused on reviews.

piodul commented on scylla/tests/retries.rs:
```diff
@@ -0,0 +1,284 @@
+    let proxy1_addr = SocketAddr::from_str(proxy1_uri.as_str()).unwrap();
+    let real2_addr = SocketAddr::from_str(real2_uri.as_str()).unwrap();
+    let proxy2_addr = SocketAddr::from_str(proxy2_uri.as_str()).unwrap();
+    let real3_addr = SocketAddr::from_str(real3_uri.as_str()).unwrap();
+    let proxy3_addr = SocketAddr::from_str(proxy3_uri.as_str()).unwrap();
+
+    let proxy = Proxy::builder()
+        .with_node(
+            Node::builder()
+                .real_address(real1_addr)
+                .proxy_address(proxy1_addr)
+                .shard_awareness(ShardAwareness::QueryNode)
+                .build(),
+        )
+        .with_node(
+            Node::builder()
+                .real_address(real2_addr)
+                .proxy_address(proxy2_addr)
+                .shard_awareness(ShardAwareness::QueryNode)
+                .build(),
+        )
+        .with_node(
+            Node::builder()
+                .real_address(real3_addr)
+                .proxy_address(proxy3_addr)
+                .shard_awareness(ShardAwareness::QueryNode)
+                .build(),
+        )
+        .build();
+
+    let running_proxy = proxy.run().with_current_subscriber().await.unwrap();
+
+    let mut translation_map = HashMap::new();
+    for (real_addr, proxy_addr) in [
+        (real1_addr, proxy1_addr),
+        (real2_addr, proxy2_addr),
+        (real3_addr, proxy3_addr),
+    ] {
+        translation_map.insert(real_addr, proxy_addr);
+    }
```
It should be possible to express this initialization as a much shorter for loop if you get rid of const generics in the Proxy and ProxyBuilder.

piodul commented on scylla/tests/retries.rs:
```diff
@@ -0,0 +1,284 @@
+        let mut q = Query::from("INSERT INTO ks2.t (a) VALUES (?)");
+        q.set_is_idempotent(true); // this is to allow speculative execution to fire
+
+        let delay_rule = RequestRule(
```
It's called `delay_rule`, but it drops requests instead of delaying them.

piodul replied:
Yes, please. I guess that "infinite delay" == "drop" in some sense, but it's still confusing. When I saw the name `delay_rule` but saw that it uses `drop_frame()` instead of the other, existing method called `delay()`, my first thought was: is it a mistake?

piodul replied:
By the way, there is another `delay_rule` in the code which does not drop or delay but rather forges a request error. That one looks even weirder.

piodul commented on scylla/tests/retries.rs:
```diff
@@ -0,0 +1,284 @@
+        // first run before any rules
+        tokio::select! {
+            res = session.query(q.clone(), (3,)) => res.unwrap(),
+            _ = tokio::time::sleep(TIMEOUT_PER_REQUEST) => panic!("Did not receive response in time."),
```
I'd get rid of this timeout, if we get stuck here then we should let the whole test time out.

piodul commented on scylla/tests/retries.rs:
```diff
@@ -0,0 +1,284 @@
+            .change_request_rules(Some(vec![delay_rule.clone()]))
+            .await;
+
+            tokio::select! {
+                res = session.query(q.clone(), (2,)) => res.unwrap(),
+                _ = tokio::time::sleep(TIMEOUT_PER_REQUEST) => panic!("Did not receive response in time."),
+            };
```
Relying on timing in tests is a bad thing. You cannot reliably check that something didn't happen just by timing out - there is always a possibility that the thing you are waiting is executing really slow.

That said, I don't see a better way to test it - after all, you do want to check here that the response does not come back. I would increase the timeout, though, due to the slowness of the machines - let's increase it to 1 second.

piodul commented on scylla/tests/retries.rs:
```diff
@@ -0,0 +1,284 @@
+
+            tokio::select! {
+                res = session.query(q.clone(), (1,)) => res.unwrap(),
+                _ = tokio::time::sleep(TIMEOUT_PER_REQUEST) => panic!("Did not receive response in time."),
```
Like previously, I think we should let the whole test time out.

piodul commented on scylla/tests/retries.rs:
```diff
@@ -0,0 +1,284 @@
+        (real2_addr, proxy2_addr),
+        (real3_addr, proxy3_addr),
+    ] {
+        translation_map.insert(real_addr, proxy_addr);
```
How about adding a method to the proxy which creates this translator/translation map? I think it will be useful in all tests that use proxy and a session.

piodul commented on scylla-proxy/src/proxy.rs:
```diff
@@ -1665,4 +1680,56 @@ mod tests {
+
+        mem::drop(driver_conn);
+        tokio::time::sleep(Duration::from_millis(10)).await;
+        assert_matches!(
+            running_proxy.sanity_check(),
+            Err(ProxyError::Worker(WorkerError::DriverDisconnected(_)))
+        );
```
This assertion depends on timing, which is bad. The test should wait until an error occurs, not just sleep and then check.

wprzytula replied:
How can I achieve that, bearing in mind how `sanity_check()` works?

piodul replied:
You could have a version of `sanity_check` which waits for an error to occur in `error_sink`. E.g. split `sanity_check` into `check_for_error` and `wait_for_error`.

wprzytula commented on scylla-proxy/src/proxy.rs:
```diff
@@ -0,0 +1,790 @@
+        let (finish_guard, finish_waiter) = mpsc::channel(1);
+
+        let (error_propagator, error_sink) = mpsc::unbounded_channel();
+        let mut doorkeepers_and_running_nodes = self.nodes.map(|node| {
```
The problem here is that I really craved for compile-time-known-length-arrays, whereas `unzip` and so on are available for iterators only (and not bare arrays), hence losing the fixed length. Conversely, `map()` applied on an arrays preserves its length, that's where the said weird tricks originate from.
If you have any ideas have to write it cleaner while still preserving the fixed length, I would strongly appreciate them.

wprzytula commented on scylla-proxy/src/actions.rs:
```diff
@@ -0,0 +1,135 @@
+                .windows(pattern.len())
+                .any(|window| *window == **pattern),
+
+            Condition::BodyContainsCaseInsensitive(pattern) => std::str::from_utf8(pattern)
+                .map(|pattern_str| {
+                    ctx.frame_body.windows(pattern.len()).any(|window| {
+                        std::str::from_utf8(window)
+                            .map(|window_str| unicase::eq_ascii(window_str, pattern_str))
+                            .unwrap_or(false)
+                    })
+                })
+                .unwrap_or(false),
```
> I don't think it is guaranteed that an uppercase and lowercase variant of a Unicode character must have the same length when encoded as UTF-8.

Right, I should have mentioned this supports ASCII characters only.

> The frame may contain the string, but it may have different length.

What do you mean? Do you assume ASCII here or not?

piodul replied:
> > The frame may contain the string, but it may have different length.
> 
> What do you mean? Do you assume ASCII here or not?

I think I just didn't read it carefully enough and missed the `unicase::eq_ascii` (btw, we have [`eq_ignore_ascii_case`](https://doc.rust-lang.org/std/primitive.str.html#method.eq_ignore_ascii_case) in the standard library).

wprzytula commented on scylla-proxy/src/proxy.rs:
```diff
@@ -66,6 +66,100 @@ impl Node {
         }
     }
+
+    pub fn builder() -> NodeWithoutRealAddress {
```
I admit: I've chosen the more complex way to do that, for the sake of learning the Rust type system better and using [the typestate pattern](http://cliffle.com/blog/rust-typestate).
The boilerplate is separated from the `meaningful` code, so its presence should not be that intrusive.

piodul replied:
> I admit: I've chosen the more complex way to do that, for the sake of learning the Rust type system better and using [the typestate pattern](http://cliffle.com/blog/rust-typestate).

I understand that the typestate pattern is interesting, but I don't think it is a good fit here - or, at least, in its current implementation.

> The boilerplate is separated from the `meaningful` code, so its presence should not be that intrusive.

Using the builder looks actually pretty nice, and it's not a problem. The problem is the builder itself. It does something very simple but at the same time is very verbose. Adding another required field to the Node essentially requires you to copy-paste the entire struct and introduce a new state in the state machine.

I'm not sure that the added compile-time safety is worth the complexity. It could be worth it if you had some complex logic that uses the builder and it would be hard to validate that all code paths always use the builder correctly. However, this is only for tests - our tests usually have very simple control flow, are short and running them exercises all of the code paths. I think validating the config in runtime will be sufficient.

wprzytula commented on scylla-proxy/src/proxy.rs:
```diff
@@ -0,0 +1,790 @@
+        );
+    }
+
+    async fn make_streams(
```
It would suggest, though, that the sole action here is `accepting` some connection, whereas this function establishes a connection to a remote node, too.

piodul replied:
OK, I won't insist on the name change.

wprzytula commented on scylla-proxy/src/proxy.rs:
```diff
@@ -0,0 +1,790 @@
+pub struct Node {
+    real_addr: SocketAddr,
+    proxy_addr: SocketAddr,
+    shard_awareness: ShardAwareness,
```
Shard awareness of the proxy implies whether it binds to a port that would preserve the advanced shard awareness properties (result of port number modulo number of shards).
With `ShardUnaware`, proxy binds to any port.
With `FixedNum(usize)`, it binds to a port that is equal to the driver's modulo the given `usize`.
With `ShardAware`, it starts with connecting to the node on any port, queries the number of shards and then establishes the proper connection from a proper port (based on the queries number of shards).

wprzytula replied:
Should I put the above explanation in the docs? I was convinced I'd already documented that well enough, hmm.

piodul replied:
I looked at the `ShardAwareness`'s definition in the commit that introduced it and found no explanation, which lead to my comment. Now I see that there are docstrings for it in the commit that introduces docstrings (I admit that I glanced over it last time).

wprzytula commented on scylla-proxy/src/proxy.rs:
```diff
@@ -0,0 +1,790 @@
+}
+
+pub struct Proxy<const CLUSTER_SIZE: usize> {
+    nodes: [InternalNode; CLUSTER_SIZE],
```
The benefit is purely educational. Will change that.

wprzytula commented on scylla-proxy/src/actions.rs:
```diff
@@ -13,22 +13,44 @@ use scylla_cql::{
     Consistency,
 };
 
+/// Specifies when an associated [Reaction] will be performed.
```
Should I refactor the code by moving the documentation to the corresponding commits or can we let it go this time?

piodul replied:
Please do. People looking at the history in git will be grateful. It will also be a good exercise in commit splitting :)

wprzytula commented:
v2: fixes related to all resolved conversations

piodul commented on scylla-proxy/src/proxy.rs:
```diff
@@ -788,3 +788,485 @@ impl ProxyWorker {
+            node1_real_addr,
+            node1_proxy_addr,
+            ShardAwareness::Unaware,
+            Some(vec![
+                RequestRule(
+                    Condition::RequestOpcode(RequestOpcode::Register),
+                    RequestReaction::forge_response(Arc::new(|RequestFrame { params, .. }| {
+                        ResponseFrame {
+                            params: params.for_response(),
+                            opcode: ResponseOpcode::Event,
+                            body: Bytes::from_static(test_msg),
+                        }
+                    })),
+                ),
+                RequestRule(
+                    Condition::BodyContainsCaseSensitive(Box::new(*this_shall_pass)),
+                    RequestReaction::noop(),
+                ),
+                RequestRule(
+                    Condition::True, // only the first matching rule is applied, so "True" covers all remaining cases
+                    RequestReaction::forge_response(Arc::new(|RequestFrame { params, .. }| {
+                        ResponseFrame {
+                            params: params.for_response(),
+                            opcode: ResponseOpcode::Ready,
+                            body: Bytes::new(),
+                        }
+                    })),
+                ),
+            ]),
```
I still stand by my preferences, but let's can keep the DSL for now. Time will tell how well it fits our needs.

piodul commented on scylla-proxy/src/proxy.rs:
```diff
@@ -0,0 +1,766 @@
+                                    let forge_action = async move {
+                                        if let Some(ref forge_action) = reaction.to_sender {
+                                            if let Some(time) = forge_action.delay {
+                                                tokio::time::sleep(time).await;
```
Do I understand correctly that this will block processing other frames until the sleep finishes?

piodul replied:
Ah, it's fixed in a later commit. Still, it's confusing when reading commit-by-commit.

piodul commented on scylla-proxy/src/proxy.rs:
```diff
@@ -0,0 +1,766 @@
+        std::mem::drop(self.finish_guard);
+    }
+
+    async fn receiver_from_driver(
```
Each of the workers have some boilerplate related to handling termination and responding with an error, I think it could be abstracted away... I tried doing it with one of the functions and it seems possible:

```rust
async fn run_until_interrupted<F, Fut>(mut self, worker_name: &'static str, f: F)
where
    F: FnOnce(SocketAddr, SocketAddr) -> Fut,
    Fut: Future<Output = Result<(), ProxyError>>,
{
    let fut = f(self.driver_addr, self.proxy_addr);

    tokio::select! {
        result = fut => {
            if let Err(err) = result {
                // error_propagator could be a field
                let _ = self.error_propagator.send(err);
            }
        }
        _ = self.terminate_notifier.recv() => {},
        _ = self.connection_close_notifier.recv() => {},
    }
    self.exit(worker_name);
}

async fn receiver_from_driver(
    self,
    mut read_half: (impl AsyncRead + Unpin),
    request_processor_tx: mpsc::UnboundedSender<RequestFrame>,
) {
    self.run_until_interrupted(
        "receiver_from_driver",
        |driver_addr, proxy_addr| async move {
            loop {
                let frame = frame::read_request_frame(&mut read_half)
                    .await
                    .map_err(|err| {
                        warn!("Request reception from {} error: {}", driver_addr, err);
                        WorkerError::DriverDisconnected(driver_addr)
                    })?;

                debug!(
                    "Intercepted Driver ({}) -> Cluster ({}) frame. opcode: {:?}.",
                    driver_addr, proxy_addr, &frame.opcode
                );
                if request_processor_tx.send(frame).is_err() {
                    warn!("request_processor had exited.");
                    return Result::<(), ProxyError>::Ok(());
                }
            }
        },
    )
    .await
}
```

This seems to reduce the amount of indents a bit.

piodul commented on scylla-proxy/src/proxy.rs:
```diff
@@ -0,0 +1,766 @@
+                request = requests_rx.recv() => {
+                    match request {
+                        Some(request) => {
+                            let ctx = EvaluationContext{
```
Looks like another bug in rustfmt - it forgot to add a space between `EvaluationContext` and `{`.

wprzytula replied:
I suspect it is due to being inside a macro (`tokio::select!`)

piodul commented:
@wprzytula in addition to the new comments, please also address my unresolved comments from previous reviews

wprzytula commented on scylla-proxy/src/errors.rs:
```diff
@@ -0,0 +1,62 @@
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum DoorkeeperError {
```
Could you write a brief example how I can use anyhow errors?

piodul replied:
Basically, you could get rid of enums in this file. Instead of using concrete error types, you use `anyhow::Error` or [`anyhow::Result<T>`](https://docs.rs/anyhow/latest/anyhow/type.Result.html) for functions that fail.

Returning errors with messages look like this:

```rust
return anyhow::anyhow!("Listen on socked {:?} failed", socket_addr);
```

Any error type converts to `anyhow::Error`. and you can attach additional context using the [`Context`](https://docs.rs/anyhow/latest/anyhow/trait.Context.html) trait:

```rust
let thing = function_that_can_fail().context("The function failed")?;
```

Refer to the documentation or `anyhow` for more details: https://docs.rs/anyhow/latest/anyhow/type.Result.html

wprzytula replied:
In some tests, I do match on errors. See `retries.rs:256` - DriverDisconnected is treated specially, as not a real error:
```
match res {
        Ok(()) => (),
        Err(ProxyError::Worker(WorkerError::DriverDisconnected(_))) => (),
        Err(err) => panic!("{}", err),
    }
```

As `thiserror`-based impl is ready and working, I would go for leaving it as is.

wprzytula commented:
> @wprzytula in addition to the new comments, please also address my unresolved comments from previous reviews

I think I've addressed all of those comments. If I'm wrong, please ping me under them.

wprzytula commented on scylla-proxy/src/proxy.rs:
```diff
@@ -0,0 +1,1716 @@
+        .await;
+    }
+
+    async fn request_processor(
```
Done.
